Guidelines for writing a connector (#1802)
ConduitIO/conduit: docs/writing-a-connector-guidelines.md

# Guidelines for writing a connector

This document describes patterns and processes that have been successfully used
in many Conduit connectors over the past few years. The document assumes basic
knowledge about how Conduit works.

## Start with the connector template

The GitHub
repository [template](https://github.com/ConduitIO/conduit-connector-template/)
for Conduit connectors provides a basic project setup that includes placeholders
for the source connector code, destination connector code, connector

> specifications, example of a configuration etc.

Reviewer: ```suggestion
specifications, and an example of a configuration, etc.
```

> Also included is a `Makefile` with commonly used target, GitHub workflow for

Reviewer: ```suggestion
Also included is a `Makefile` with commonly used targets. And some GitHub workflows for
```

in a better connector design and better work organization.

Some questions that typically need to be answered:

1. **How is the data organized in the system?**

   The way data is organized in the system will affect how the connector is
   configured (e.g. what data is a user able to read or write to) and how the
   hierarchy is mapped to records and collections in Conduit.
2. **What parameters are needed to connect to the system?**

   We generally recommend using connection strings/URIs, if available. The
   reason for this is that modifying a connector's parameters is a matter of
   changing an existing string, and not separate configuration parameters in a
   connector.
3. **What APIs or drivers are available?**

   The choice is influenced by a number of factors, such as:
   - Is there a driver available? If a public API and a driver are available, we
     recommend using a driver, since it's a higher level of abstraction.
   - In which language should the connector be written?
   - If the language of choice is Go, is a pure Go driver available? A CGo
     driver may make the usage and deployment of a connector more complex.

>    - Does the driver expose all the functionality needed?

Reviewer: ```suggestion
   - Does the driver expose all the functionalities needed?
```

   - How mature is the driver?
4. **What authentication methods should be supported?**

   If multiple authentication methods are available, then a decision needs to be
   made about the methods that should be implemented first. Additionally, it
   should be understood how expired credentials should be handled. For example,
   a connector won't be able to handle an expired password, but a token can
   sometimes be refreshed.
5. **Can the connector be isolated from other clients of the system?**

   In some cases a connector, as a client using a system, might affect other
   clients, for example when getting messages from a message broker's queue, the
   message will be delivered to the connector, while other clients might expect
   the message.
6. **Are the source/destination specific features supported?**

   Source and destination connectors may have specific requirements (some of
   them are outlined in later sections). When researching, attention should be
   paid if those requirements can be fulfilled.

> ## Development
> 
> ## Familiarize with the Connector SDK/connector protocol

Reviewer 2: Something doesn't seem right here, I suppose the second header would need to be nested below the first?

```suggestion
## Development

### Familiarize with the Connector SDK/connector protocol
```

Conduit's Connector SDK is the Go software development kit for implementing a
connector for Conduit. If you want to implement a connector in another language

> please have a look at

Reviewer: ```suggestion
please check
```

#### Deciding how should a record position look like

Every record read by a source connector has
a [position](https://conduit.io/docs/features/opencdc-record#fields) attached.
If a connector is stopped after that record is read, its position will be given
to the connector's `Open()` method the next time it starts. Hence, a position
needs to contain enough information for a source connector to resume reading
records from where it exactly stopped.

A position is a slice of bytes that can represent any data structure. In Conduit
connectors, it's common to see that a position is actually a `struct`, that's
marshalled into a JSON string.

#### Implementing snapshots

Firstly, it should be clarified if supporting snapshots is a requirement or if
it's possible to do at all. If a connector is required to support snapshots,
then it's recommended to make it possible to turn off snapshots through the
connector configuration.

The following things need to be taken into account when implementing a snapshot
procedure:

- The snapshot needs to be consistent.
- The set of the existing data can be quite large.
- Restarting a connector during a snapshot should **not** require re-reading all

>   the data again. This is because it in some destination connectors it may cause

Reviewer: ```suggestion
  the data again. This is because in some destination connectors it may cause
```
